File: JavaIntroNeuralNetworkEdition2/src/com/heatonresearch/book/introneuralnet/neural/feedforward/train/backpropagation/Backpropagation.java
/**
 * Introduction to Neural Networks with Java, 2nd Edition
 * Copyright 2008 by Heaton Research, Inc. 
 * http://www.heatonresearch.com/books/java-neural-2/
 * 
 * ISBN13: 978-1-60439-008-7  	 
 * ISBN:   1-60439-008-5
 *   
 * This class is released under the:
 * GNU Lesser General Public License (LGPL)
 * http://www.gnu.org/copyleft/lesser.html
 */
package com.heatonresearch.book.introneuralnet.neural.feedforward.train.backpropagation;

import java.util.HashMap;
import java.util.Map;

import com.heatonresearch.book.introneuralnet.neural.exception.NeuralNetworkError;
import com.heatonresearch.book.introneuralnet.neural.feedforward.FeedforwardLayer;
import com.heatonresearch.book.introneuralnet.neural.feedforward.FeedforwardNetwork;
import com.heatonresearch.book.introneuralnet.neural.feedforward.train.Train;

/**
 * Backpropagation: This class implements a backpropagation 
 * training algorithm for feed forward neural networks.
 * It is used in the same manner as any other training class 
 * that implements the Train interface.
 * 
 * Backpropagation is a common neural network training algorithm.
 * It works by analyzing the error of the output of the neural 
 * network.  Each neuron in the output layer's contribution, according
 * to weight, to this error is determined.  These weights are 
 * then adjusted to minimize this error.  This process continues
 * working its way backwards through the layers of the neural
 * network.
 * 
 * This implementation of the backpropagation algorithm uses both
 * momentum and a learning rate.  The learning rate specifies the 
 * degree to which the weight matrixes will be modified through
 * each iteration.  The momentum specifies how much the previous
 * learning iteration affects the current.  To use no momentum
 * at all specify zero.
 * 
 * @author Jeff Heaton
 * @version 2.1
 */
public class Backpropagation implements Train {
	/**
	 * The error from the last iteration.
	 */
	private double error;

	/**
	 * The learning rate. This is the degree to which the deltas will affect the
	 * current network.
	 */
	private final double learnRate;

	/**
	 * The momentum, this is the degree to which the previous training cycle
	 * affects the current one.
	 */
	private final double momentum;

	/**
	 * THe network that is being trained.
	 */
	private final FeedforwardNetwork network;

	/**
	 * A map between neural network layers and the corresponding
	 * BackpropagationLayer.
	 */
	private final Map<FeedforwardLayer, BackpropagationLayer> layerMap = new HashMap<FeedforwardLayer, BackpropagationLayer>();

	/**
	 * Input patterns to train with.
	 */
	private final double input[][];

	/**
	 * The ideal output for each of the input patterns.
	 */
	private final double ideal[][];

	/**
	 * 
	 * @param network
	 * @param input
	 * @param ideal
	 * @param learnRate
	 *            The rate at which the weight matrix will be adjusted based on
	 *            learning.
	 * @param momentum
	 *            The influence that previous iteration's training deltas will
	 *            have on the current iteration.
	 */
	public Backpropagation(final FeedforwardNetwork network,
			final double input[][], final double ideal[][],
			final double learnRate, final double momentum) {
		this.network = network;
		this.learnRate = learnRate;
		this.momentum = momentum;
		this.input = input;
		this.ideal = ideal;

		for (final FeedforwardLayer layer : network.getLayers()) {
			final BackpropagationLayer bpl = new BackpropagationLayer(this,
					layer);
			this.layerMap.put(layer, bpl);
		}
	}

	/**
	 * Calculate the error for the recognition just done.
	 * 
	 * @param ideal
	 *            What the output neurons should have yielded.
	 */
	public void calcError(final double ideal[]) {

		if (ideal.length != this.network.getOutputLayer().getNeuronCount()) {
			throw new NeuralNetworkError(
					"Size mismatch: Can't calcError for ideal input size="
							+ ideal.length + " for output layer size="
							+ this.network.getOutputLayer().getNeuronCount());
		}

		// clear out all previous error data
		for (final FeedforwardLayer layer : this.network.getLayers()) {
			getBackpropagationLayer(layer).clearError();
		}

		for (int i = this.network.getLayers().size() - 1; i >= 0; i--) {
			final FeedforwardLayer layer = this.network.getLayers().get(i);
			if (layer.isOutput()) {

				getBackpropagationLayer(layer).calcError(ideal);
			} else {
				getBackpropagationLayer(layer).calcError();
			}
		}
	}

	/**
	 * Get the BackpropagationLayer that corresponds to the specified layer.
	 * @param layer The specified layer.
	 * @return The BackpropagationLayer that corresponds to the specified layer.
	 */
	public BackpropagationLayer getBackpropagationLayer(
			final FeedforwardLayer layer) {
		final BackpropagationLayer result = this.layerMap.get(layer);

		if (result == null) {
			throw new NeuralNetworkError(
					"Layer unknown to backpropagation trainer, was a layer added after training begain?");
		}

		return result;
	}

	/**
	 * Returns the root mean square error for a complete training set.
	 * 
	 * @param len
	 *            The length of a complete training set.
	 * @return The current error for the neural network.
	 */
	public double getError() {
		return this.error;
	}

	/**
	 * Get the current best neural network.
	 * @return The current best neural network.
	 */
	public FeedforwardNetwork getNetwork() {
		return this.network;
	}

	/**
	 * Perform one iteration of training.
	 */
	public void iteration() {

		for (int j = 0; j < this.input.length; j++) {
			this.network.computeOutputs(this.input[j]);
			calcError(this.ideal[j]);
		}
		learn();
		
		this.error = this.network.calculateError(this.input, this.ideal);
	}

	/**
	 * Modify the weight matrix and thresholds based on the last call to
	 * calcError.
	 */
	public void learn() {

		for (final FeedforwardLayer layer : this.network.getLayers()) {
			getBackpropagationLayer(layer).learn(this.learnRate, this.momentum);
		}

	}
}

File: HeatonResearchNeural/Feedforward/Train/Backpropagation/Backpropagation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using HeatonResearchNeural.Exception;

namespace HeatonResearchNeural.Feedforward.Train.Backpropagation
{
    public class Backpropagation : Train
    {
        /// <summary>
        /// Get the current best neural network.
        /// </summary>
        public FeedforwardNetwork Network
        {
            get
            {
                return this.network;
            }
        }


        /// <summary>
        /// Returns the root mean square error for a complete training set.
        /// </summary>

        public double Error
        {
            get
            {
                return this.error;
            }
        }



        /// <summary>
        /// The error from the last iteration.
        /// </summary>
        private double error;


        /// <summary>
        /// The learning rate. This is the degree to which the deltas will affect the
        /// current network.
        /// </summary>
        private double learnRate;


        /// <summary>
        /// The momentum, this is the degree to which the previous training cycle
        /// affects the current one.
        /// </summary>
        private double momentum;

        /// <summary>
        /// The network that is being trained.
        /// </summary>
        private FeedforwardNetwork network;

        /// <summary>
        /// A map between neural network layers and the corresponding
        /// BackpropagationLayer.
        /// </summary>
        private IDictionary<FeedforwardLayer, BackpropagationLayer> layerMap = new Dictionary<FeedforwardLayer, BackpropagationLayer>();

        /// <summary>
        /// Input patterns to train with.
        /// </summary>
        private double[][] input;

        /// <summary>
        /// The ideal output for each of the input patterns.
        /// </summary>
        private double[][] ideal;

        /// <summary>
        /// The rate at which the weight matrix will be adjusted based on
        /// learning.
        /// </summary>
        /// <param name="network">The neural network to be trained.</param>
        /// <param name="input">The input values to the neural network.</param>
        /// <param name="ideal">The ideal values expected from the neural network.</param>
        /// <param name="learnRate">The learning rate, how fast to modify neural network values.</param>
        /// <param name="momentum">The momentum, how much to use the previous training iteration for the current.</param>
        public Backpropagation(FeedforwardNetwork network,
                 double[][] input, double[][] ideal,
                 double learnRate, double momentum)
        {
            this.network = network;
            this.learnRate = learnRate;
            this.momentum = momentum;
            this.input = input;
            this.ideal = ideal;

            foreach (FeedforwardLayer layer in network.Layers)
            {
                BackpropagationLayer bpl = new BackpropagationLayer(this,
                       layer);
                this.layerMap.Add(layer, bpl);
            }
        }

        /// <summary>
        /// Calculate the error for the recognition just done.
        /// </summary>
        /// <param name="ideal">What the output neurons should have yielded.</param>
        public void CalcError(double[] ideal)
        {

            if (ideal.Length != this.network.OutputLayer.NeuronCount)
            {
                throw new NeuralNetworkError(
                        "Size mismatch: Can't calcError for ideal input size="
                                + ideal.Length + " for output layer size="
                                + this.network.OutputLayer.NeuronCount);
            }

            // clear out all previous error data
            foreach (FeedforwardLayer layer in this.network.Layers)
            {
                GetBackpropagationLayer(layer).ClearError();
            }

            for (int i = this.network.Layers.Count - 1; i >= 0; i--)
            {
                FeedforwardLayer layer = this.network.Layers[i];
                if (layer.IsOutput())
                {

                    GetBackpropagationLayer(layer).CalcError(ideal);
                }
                else
                {
                    GetBackpropagationLayer(layer).CalcError();
                }
            }
        }

        /// <summary>
        /// Get the BackpropagationLayer that corresponds to the specified layer.
        /// </summary>
        /// <param name="layer">The specified layer.</param>
        /// <returns>The BackpropagationLayer that corresponds to the specified layer.</returns>
        public BackpropagationLayer GetBackpropagationLayer(
                 FeedforwardLayer layer)
        {
            BackpropagationLayer result = this.layerMap[layer];

            if (result == null)
            {
                throw new NeuralNetworkError(
                        "Layer unknown to backpropagation trainer, was a layer added after training begain?");
            }

            return result;
        }





        /// <summary>
        /// Perform one iteration of training.
        /// </summary>
        public void Iteration()
        {

            for (int j = 0; j < this.input.Length; j++)
            {
                this.network.ComputeOutputs(this.input[j]);
                CalcError(this.ideal[j]);
            }
            Learn();

            this.error = this.network.CalculateError(this.input, this.ideal);
        }

        /// <summary>
        /// Modify the weight matrix and thresholds based on the last call to
        /// calcError.
        /// </summary>
        public void Learn()
        {

            foreach (FeedforwardLayer layer in this.network.Layers)
            {
                GetBackpropagationLayer(layer).Learn(this.learnRate, this.momentum);
            }

        }

    }
}
